### PR TITLE
Editor: Make sure `resetEditorBlocks` is synchronous.

### LINKED ISSUE #21804: Controlled Inner Blocks Clear On Mount
Something has regressed with controlled inner blocks where initial values are overwritten with an empty array shortly after mount.

This is noticeable when using the Template Part block placeholder to insert existing template parts. The preview works, but the block is empty after insertion.

### PR DESCRIPTION
**Pull request by epiqueras**
Fixes #21804

## Description

#21467 introduced a resolver for `getEditedEntityRecord` which made `resetEditorBlocks` asynchronous. This sometimes makes previous calls finish executing after newer ones, essentially undoing the editor to a stale block tree.

I added comments to the relevant code to explain:

```js
export function* resetEditorBlocks( blocks, options = {} ) {
	const {
		__unstableShouldCreateUndoLevel,
		selectionStart,
		selectionEnd,
	} = options;
	const edits = { blocks, selectionStart, selectionEnd };
	// First Run: `[ core/template-part: [] ], undefined`.
	// Second Run: `[ core/template-part: [ ...initialBlocks ] ], false`.
	console.log( blocks, __unstableShouldCreateUndoLevel );
	if ( __unstableShouldCreateUndoLevel !== false ) {
		const { id, type } = yield select( STORE_KEY, 'getCurrentPost' );
		const noChange =
			( yield select(
				'core',
				'getEditedEntityRecord',
				'postType',
				type,
				id
			) ).blocks === edits.blocks;
		// First Run: `false`.
		// Second Run: Doesn't reach this code.
		// getEditedEntityRecord having a resolver makes
		// the execution from the first run reach this point
		// after the second run finishes. This overwrites the
		// blocks with a stale value. A stale value without
		// template part inner blocks which causes the issue
		// described here.
		console.log( noChange );
		if ( noChange ) {
			return yield dispatch(
				'core',
				'__unstableCreateUndoLevel',
				'postType',
				type,
				id
			);
		}
		// We create a new function here on every persistent edit
		// to make sure the edit makes the post dirty and creates
		// a new undo level.
		edits.content = ( { blocks: blocksForSerialization = [] } ) =>
			serializeBlocks( blocksForSerialization );
	}
	yield* editPost( edits );
}
```

This PR fixes the issue by using a synchronous select data control like we used to have before.

But, I think that this problem could surface elsewhere, so the long-term solution should be more general. We should think about canceling stale action execution. That also raises the questions of how to define and continue with any cleanup operations, though.

## How to test this?

Verify that #21804 is no longer an issue; that you can insert existing template parts on the full site editing experiment as expected.

## Types of Changes

*Bug Fix:* `resetEditorBlocks` will now execute sequentially, avoiding an issue where older executions would overwrite newer ones.
*New Feature:* `@wordpress/data-controls` now has a synchronous version of `select`, `syncSelect`, which does not wait for resolvers.

## Checklist:

- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by aduth** [comment]
>Verify that #21804 is no longer an issue; that you can insert existing template parts on the full site editing experiment as expected.

Since it took me a while to figure out the testing instructions:

1. (Prerequisite) Activate and save "Enable Full Site Editing" from Gutenberg > Experiments
2. Navigate to Posts > Add New
3. Insert a "Template Part" block
4. Enter "header" for Slug and "twentytwenty" for Theme
5. Press "Choose"
6. Note the text "Header Template Part" appears
   - In `master`, it's wrongly empty prompt "Start writing or type / to choose block"

**Comment by aduth** [comment]
>But, I think that this problem could surface elsewhere, so the long-term solution should be more general. We should think about canceling stale action execution. That also raises the questions of how to define and continue with any cleanup operations, though.

Do we want a new public API member `syncSelect` then? Should it be `__unstable`? Or local to the `editor` package like we do with [some similar controls](https://github.com/WordPress/gutenberg/blob/aa44b76ee4f3e39047dc2937f861cb256094d279/packages/core-data/src/controls.js#L20-L34)?† Or is there valid use for this beyond addressing the immediate issue here?

_† I believe this example isn't done this way on purpose, merely that it was implemented before the package was generalized._

**Comment by aduth** [comment]
Do we risk any inaccurate value from `getEditedEntityRecord` by not awaiting the resolver? I guess if it wasn't awaiting anything before, this would restore that previous behavior? (Which doesn't speak to whether or not it _should_ be waiting for e.g. `getEntityRecord`).

**Comment by aduth** [comment]
Not sure if this was introduced here or in #21804, or has always been an issue, but I see a lot of duplicate toolbar and inspector controls for a freshly-inserted Template Part block. Plus, when trying to interact with the block, it locks up my editor.

image

**Comment by aduth** [comment]
> > But, I think that this problem could surface elsewhere, so the long-term solution should be more general. We should think about canceling stale action execution. That also raises the questions of how to define and continue with any cleanup operations, though.
> 
> Do we want a new public API member `syncSelect` then? Should it be `__unstable`? Or local to the `editor` package like we do with [some similar controls](https://github.com/WordPress/gutenberg/blob/aa44b76ee4f3e39047dc2937f861cb256094d279/packages/core-data/src/controls.js#L20-L34)?† Or is there valid use for this beyond addressing the immediate issue here?

On the topic, while I don't think it's inaccurate to name it as such, the specific use of "sync" leads me to think of it differently, in-fact more like how it actually behaves today. When I see the name, I consider it like [Node's `readFileSync`](https://nodejs.org/api/fs.html#fs_fs_readfilesync_path_options), where it would wait before proceeding. The actual behavior is more like _don't wait, just use state as it exists at the moment_.

**Review by aduth — APPROVED** [review]
I'd probably prefer to hide `syncSelect` from the public API in its current form, but otherwise it appears to fix the issue in my testing.

Separately, we should consider some automated testing coverage for this behavior, especially since it's fixing a regression.

**Comment by epiqueras** [comment]
>Do we risk any inaccurate value from getEditedEntityRecord by not awaiting the resolver? I guess if it wasn't awaiting anything before, this would restore that previous behavior? (Which doesn't speak to whether or not it should be waiting for e.g. getEntityRecord).

No, it should be fine.

>Not sure if this was introduced here or in #21804, or if it's always been an issue, but I see a lot of duplicate toolbar and inspector controls for a freshly-inserted Template Part block. Plus, when trying to interact with the block, it locks up my editor.

Yes, this is because they are the same template parts, so they share block instances/IDs. We need to fix this.

>On the topic, while I don't think it's inaccurate to name it as such, the specific use of "sync" leads me to think of it differently, in-fact more like how it actually behaves today. When I see the name, I consider it like Node's readFileSync, where it would wait before proceeding. The actual behavior is more like don't wait, just use state as it exists at the moment.

What would you call it? `noResolverSelect` seemed too wordy.

>I'd probably prefer to hide syncSelect from the public API in its current form, but otherwise it appears to fix the issue in my testing.

Agreed, I'll make it unstable for now. But, we need to arrive at a long-term solution because this will also be a problem for third parties.

>Separately, we should consider some automated testing coverage for this behavior, especially since it's fixing a regression.

Yes, I want to start with e2e tests for template parts, which should indirectly test a lot of the new behaviors.

**Comment by aduth** [comment]
> What would you call it? `noResolverSelect` seemed too wordy.

I guess it doesn't really matter much if it's going to be internal / unstable. I don't really have great name alternatives in mind, maybe something reflecting the "immediacy" or "current state" aspects of the behavior. I also kinda wished it would be an option of the existing `select`, but given the variadic nature of the existing function, we can't easily add an "options" argument. I'm not too worried about it.

**Comment by aduth** [comment]
> > I'd probably prefer to hide syncSelect from the public API in its current form, but otherwise it appears to fix the issue in my testing.
> 
> Agreed, I'll make it unstable for now. But, we need to arrive at a long-term solution because this will also be a problem for third parties.

Yeah, but unless I'm mistaken, we don't want anything like _this_ to be the end solution, which also makes it a perfect use-case for `__unstable`. The fact that this became an issue isn't a result of us _explicitly wanting_ to not wait for resolvers attached to the selectors called, it's just that we unintentionally depended on that behavior. I think what you've suggested about cancelling or queueing these actions / resolvers to occur in a predictable way would be much more durable, and I'd doubt it would require the `syncSelect` introduced here.

**Comment by epiqueras** [comment]
>I guess it doesn't really matter much if it's going to be internal / unstable. I don't really have great name alternatives in mind, maybe something reflecting the "immediacy" or "current state" aspects of the behavior. I also kinda wished it would be an option of the existing select, but given the variadic nature of the existing function, we can't easily add an "options" argument. I'm not too worried about changing anything.

Yeah, I reached for that first as well.

>Yeah, but unless I'm mistaken, we don't want anything like this to be the end solution, which also makes it a perfect use-case for __unstable. The fact that this became an issue isn't a result of us explicitly wanting to not wait for resolvers attached to the selectors called, it's just that we unintentionally depended on that behavior. I think what you've suggested about cancelling or queueing these actions / resolvers to occur in a predictable way would be much more durable, and I'd doubt it would require the syncSelect introduced here.

Exactly, that's what I was referring too.
